Frida persistent

WorksButNotTested opened the pull request:
I have now added support on x64 for persistent mode (function call only so far), in-memory test cases and complog.
Grateful for any feedback.

vanhauser-thc commented:
This is for the first round:

1. please merge from dev backwards and do `make code-format` then push the changes. this will properly format your code :)

2. please move all test files like frida_parallel.sh etc. into the test/ subfolder - or delete them.
also put -D in all the afl-fuzz lines, this eliminates most of the randomness and is better for testing.
Also maybe put the test cases of the Makefile into the test/ subfolder so the main Makefile is easier to read.
And put a newline at the end of all files

too tired today to thoroughly look through the code now :)

domenukk commented on frida_mode/Makefile:
```diff
@@ -1,13 +1,17 @@
 BUILD_DIR:=$(PWD)build/
 CFLAGS+=-fPIC -D_GNU_SOURCE
+QEMU_INC_DIR:=$(PARENT)qemu_mode/qemuafl/qemuafl/
+QEMU_INC_API:=$(QEMU_INC_DIR)api.h
```
Frida mode should probably not depend on qemuafl being checked out?

WorksButNotTested replied:
I'm hoping to share much of the code for qemu asan down the line. There enough similarity that it doesn't make sense to re-implement it all. So I figured a dependency on QEMU mode was preferable to duplication?

vanhauser-thc replied:
The thing is that qemuafl is not necessarily checked out. I think it is fine if you copy the file and put it into the frida_mode include directory

domenukk replied:
Or we can pull it up to aflplusplus directly, and include it like we do with `types.h` and `config.h`?

WorksButNotTested replied:
Turns out it's only a single structure I need for the CPU regs, so I have just duplicated it. I only need it as I have reused the `utils/aflpp_driver/aflpp_qemu_driver_hook.so` and so need to pass the context in the same format.

Down the line, its [this](https://github.com/AFLplusplus/AFLplusplus/tree/stable/qemu_mode/libqasan) code I expect to be using, so that's not in qemuafl and so doesn't have the same problems.

domenukk commented on frida_mode/frida_parallel.sh:
```diff
@@ -0,0 +1,33 @@
+
+rm -rf $FRIDA_OUT
+
+tmux new-window -n:master  "AFL_FRIDA_PERSISTENT_ADDR=$AFL_FRIDA_PERSISTENT_ADDR AFL_FRIDA_INST_RANGES=$AFL_FRIDA_INST_RANGES AFL_FRIDA_INST_STRICT=1 $PARENT/afl-fuzz -O -V60 -i $TEST_DATA_DIR -o $FRIDA_OUT -D master  -- $TEST_BIN @@"
```
`main` and `secondary` nodes, please :)

WorksButNotTested replied:
These scripts are gone now.

domenukk commented on frida_mode/frida_parallel_hook.sh:
```diff
@@ -0,0 +1,36 @@
+
+rm -rf $FRIDA_OUT
+
+tmux new-window -n:master  "AFL_FRIDA_PERSISTENT_HOOK=$AFL_FRIDA_PERSISTENT_HOOK AFL_FRIDA_PERSISTENT_ADDR=$AFL_FRIDA_PERSISTENT_ADDR AFL_FRIDA_INST_RANGES=$AFL_FRIDA_INST_RANGES AFL_FRIDA_INST_STRICT=1 $PARENT/afl-fuzz -O -V60 -i $TEST_DATA_DIR -o $FRIDA_OUT -D master  -- $TEST_BIN $AFLPP_DRIVER_DUMMY_INPUT"
```
Same

WorksButNotTested replied:
These scripts are gone now.

domenukk commented on frida_mode/qemu_parallel.sh:
```diff
@@ -0,0 +1,33 @@
+
+rm -rf $QEMU_OUT
+
+tmux new-window -n:master  "AFL_QEMU_PERSISTENT_ADDR=$AFL_QEMU_PERSISTENT_ADDR AFL_QEMU_INST_RANGES=$AFL_QEMU_INST_RANGES AFL_QEMU_PERSISTENT_GPR=1 $PARENT/afl-fuzz -V60 -Q -i $TEST_DATA_DIR -o $QEMU_OUT -D master  -- $TEST_BIN @@"
```
3rd time.
In any case, do we want these scripts?

WorksButNotTested replied:
These scripts are gone now.

Commit:
Added x64 support for persistent mode (function call only), in-memory teest cases and complog

domenukk commented on frida_mode/src/complog/complog.c:
```diff
@@ -0,0 +1,72 @@
+static gboolean complog_overlaps(GumAddress base1, GumAddress limit1,
+                                 GumAddress base2, GumAddress limit2) {
+
+  if (base1 < base2) return false;
```
Could be simplified?

WorksButNotTested replied:
Done.

domenukk commented on frida_mode/src/complog/complog_x64.c:
```diff
@@ -0,0 +1,367 @@
+  switch (ctx->type) {
+
+    case X86_OP_REG:
+      return complog_read_reg(context, ctx->reg);
```
What about non-64 byte compares?

WorksButNotTested replied:
The function `complog_read_reg` deals with comparisons of different sizes. Although a 64bit type is used to represent the value, it is masked and shifted as necessary.

domenukk commented on frida_mode/src/complog/complog_x64.c:
```diff
@@ -0,0 +1,367 @@
+static void complog_instrument_call_put_callout(GumStalkerIterator *iterator,
+                                                cs_x86_op *         operand) {
+
+  complog_ctx_t *ctx = g_malloc(sizeof(complog_ctx_t));
```
unchecked malloc?

WorksButNotTested replied:
Fixed.

domenukk commented on frida_mode/src/complog/complog_x64.c:
```diff
@@ -0,0 +1,367 @@
+                                                   cs_x86_op *         operand1,
+                                                   cs_x86_op *operand2) {
+
+  complog_pair_ctx_t *ctx = g_malloc(sizeof(complog_pair_ctx_t));
```
unchecked

WorksButNotTested replied:
Fixed

domenukk commented on frida_mode/src/complog/complog_x64.c:
```diff
@@ -0,0 +1,367 @@
+  if (operand1->type == X86_OP_INVALID) return;
+  if (operand2->type == X86_OP_INVALID) return;
+
+  if (operand1->type == X86_OP_MEM && operand1->mem.segment != X86_REG_INVALID)
```
I personally perfer brakets around conditions, especially if they are multiline :)

WorksButNotTested replied:
Done

Commit:
Review changes, fix NeverZero and code to parse the .text section of the main executable. Excluded ranges TBC

Commit:
Various minor fixes and finished support for AFL_INST_LIBS

WorksButNotTested commented:
I have now added support for `AFL_INST_LIBS`, restructured things to get all the test code out of the way and updated the docs. Let me know what you think.

vanhauser-thc commented on frida_mode/Makefile:
```diff
@@ -1,86 +1,58 @@
-VERSION=14.2.13
-GUM_DEVKIT_FILENAME=frida-gum-devkit-$(VERSION)-$(OS)-$(ARCH).tar.xz
-GUM_DEVKIT_URL="https://github.com/frida/frida/releases/download/$(VERSION)/$(GUM_DEVKIT_FILENAME)"
+GUM_DEVKIT_VERSION=14.2.13
```
current is 14.2.17
do you want to pin it to that specific version? or update the version from time to time?
you could also get the most current version automatically and if in the future this fails then its pinned by hand (or fixed)?

WorksButNotTested replied:
I have bumped the version to the latest. I think I would rather hard-code the version so we have the opportunity to test new versions rather than potentially hit issues in the unlikely event that a new FRIDA release breaks something. Given the makefile is now very simple, I would expect someone wanting to use a newer version would be easily able to do so. In any case, I expect that people hacking on this would actually prefer to build FRIDA from source, but I didn't want to add it as submodules and incur all of their dependencies and additional compilation time.

vanhauser-thc commented on utils/qemu_persistent_hook/read_into_rdi.c:
```diff
@@ -5,7 +5,7 @@
 void afl_persistent_hook(struct x86_64_regs *regs, uint64_t guest_base,
                          uint8_t *input_buf, uint32_t input_buf_len) {
-
+\
```
wrong blackslash

WorksButNotTested replied:
Removed, it looks like the auto formatting might be adding this in?

vanhauser-thc commented on utils/aflpp_driver/aflpp_qemu_driver_hook.c:
```diff
@@ -5,7 +5,7 @@
 void afl_persistent_hook(struct x86_64_regs *regs, uint64_t guest_base,
                          uint8_t *input_buf, uint32_t input_buf_len) {
-
+\
```
wrong blackslash

WorksButNotTested replied:
Removed, it looks like the auto formatting might be adding this in?

vanhauser-thc commented on src/afl-fuzz-init.c:
```diff
@@ -2774,6 +2774,14 @@ void check_binary(afl_state_t *afl, u8 *fname) {
 
+  } else if (getenv("AFL_FRIDA_PERSISTENT_ADDR")) {
+
+    OKF(cPIN "FRIDA Persistent mode configuration options detected.");
```
does not need cPIN IMHO

WorksButNotTested replied:
Removed

vanhauser-thc commented:
please install clang-format-11, do a `make code-format` and push the changes.

besides the comments it looks good in my opinion.

domenukk commented on frida_mode/Makefile:
```diff
@@ -1,86 +1,58 @@
 PWD:=$(shell pwd)/
```
The file should probably be a `GNUMakefile` (unless it really doesn't use GNU), we usually print a little info for non-gnu systems.

WorksButNotTested replied:
I've moved everything to using GNUmakefile now with a Makefile wrapper.

domenukk commented on frida_mode/README.md:
```diff
@@ -10,23 +10,23 @@ a small harness around their target code of interest, FRIDA mode instead takes a
+  | -------------------------|:----------:|:---------------------------------------:|
+  | NeverZero                |     x      |                                         |
+  | Persistent Mode          |     x      | (x64 only)(Only on function boundaries) |  
+  | LAF-Intel / CompCov      |     -      | (Superseded by CmpLog)                  |  
```
Wouldn't it be cool (and not to hard) to support both, for different use-cases?

vanhauser-thc replied:
compcov is very different to implement, much more work and less often effective than cmplog.
so putting compcov on the todo list but putting it somewhere on the bottom is the right thing IMHO :)

WorksButNotTested replied:
Happy to add it somewhere down the line if there's a need. When I was discussing with @vanhauser-thc, he seemed to say that CompLog was much better? If there are use-cases where CompCov is better, or CompLog unsuitable, then I am happy to re-evaluate? CompLog support shouldn't be too hard, but it will take some time and would need implementing for each architecture.

domenukk replied:
if you have decoded the instruction already, it should be straight forward to emit compcov coverage at that spot in frida.
I agree complog is better, though. Ignore it for now :)

domenukk commented on frida_mode/include/stalker.h:
```diff
@@ -0,0 +1,8 @@
+#include "frida-gum.h"
+
+void        stalker_init();
```
This should be
```c
void stalker_init(void);
```

(also true for the following lines)

WorksButNotTested replied:
I have fixed this issue for all sources and headers now.

domenukk commented on frida_mode/include/ranges.h:
```diff
@@ -1,6 +1,6 @@
 #include "frida-gum.h"
 
-void ranges_init(GumStalker *stalker);
+void ranges_init();
```
should be `void ranges_init(void);`

WorksButNotTested replied:
I have fixed this issue for all sources and headers now.

domenukk commented on frida_mode/src/complog/complog.c:
```diff
@@ -0,0 +1,72 @@
+
+}
+
+void complog_init() {
```
`void cmplog_init(void) {`

WorksButNotTested replied:
I have fixed this issue for all sources and headers now.

domenukk commented on frida_mode/src/complog/complog.c:
```diff
@@ -0,0 +1,72 @@
+
+}
+
+gint complog_sort(gconstpointer a, gconstpointer b) {
```
I guess this could be `static`?

WorksButNotTested replied:
Fixed

domenukk commented on frida_mode/src/complog/complog_x64.c:
```diff
@@ -0,0 +1,427 @@
+    case X86_REG_EDI:
+      return ctx->rdi & GUM_INT32_MASK;
+
+    case X86_REG_R8D:
```
This duplicated code could be simplified/combined with `#define`s

WorksButNotTested replied:
Fixed.

domenukk commented on frida_mode/src/instrument/instrument.c:
```diff
@@ -0,0 +1,145 @@
+
+uint64_t __thread previous_pc = 0;
+
+static void on_basic_block(GumCpuContext *context, gpointer user_data) {
```
you can try marking this with `__attribute__ ((hot))` and see if it improves speed (or not, who knows)

WorksButNotTested replied:
It can't do any harm, so I've added it in. Should only take effect on x86 and aarch32 for now, or with AFL_FRIDA_NO_OPTIMISE

domenukk commented on frida_mode/src/instrument/instrument.c:
```diff
@@ -0,0 +1,145 @@
+  guint64     current_pc = (guint64)user_data;
+  uint8_t *   cursor;
+  uint64_t    value;
+  if (tracing) {
```
Maybe make this `unlikely(tracing)` if you have unlikely available, or define it?

WorksButNotTested replied:
Fixed

domenukk commented on frida_mode/src/instrument/instrument.c:
```diff
@@ -0,0 +1,145 @@
+  cursor = &__afl_area_ptr[current_pc ^ previous_pc];
+  value = *cursor;
+
+  if (value == 0xffffffffffffffff)
```
This can never be true, as `*cursor` is of (unsigned) type `uint8_t`. Max value is 0xff.
Tested it here: https://godbolt.org/z/498zx1K87

P.S.: I personally prefer brackets around `if` statements, especially with subsequent `else`

WorksButNotTested replied:
Good spot. Weird the compiler didn't warn me! Fixed

Commit:
Review changes

WorksButNotTested commented:
> please install clang-format-11, do a `make code-format` and push the changes.
> 
> besides the comments it looks good in my opinion.

I've run the make code-format now. I also have a `make format` target of my own which just does the frida mode code. Like I said though, it seems keen to introduce those weird backslashes?

vanhauser-thc commented:
LGTM
